esp-wifi/src/wifi/mod.rs

```diff
@@ -645,23 +645,27 @@ unsafe extern "C" fn recv_cb(
     len: u16,
     eb: *mut c_types::c_void,
 ) -> esp_err_t {
-    critical_section::with(|cs| {
+    let result = critical_section::with(|cs| {
         let packet = EspWifiPacketBuffer { buffer, len, eb };
 
-        match DATA_QUEUE_RX.borrow_ref_mut(cs).enqueue(packet) {
-            Ok(_) => {
-                #[cfg(feature = "embassy-net")]
-                embassy::RECEIVE_WAKER.wake();
+        DATA_QUEUE_RX.borrow_ref_mut(cs).enqueue(packet)
+    });
 
-                include::ESP_OK as esp_err_t
-            }
+    // We must handle the result outside of the critical section because
+    // EspWifiPacketBuffer::drop must not be called in a critical section
```

Comment:
Do we know why? The (albeit brief) docs for `esp_wifi_internal_free_rx_buffer` don't mention this.

Reply:
See my stack trace. It tries to take the mutex that loops because we're calling `esp_wifi_internal_free_rx_buffer ` in a critical section

Reply:
Ah, I see, sorry for missing that. Could you expand the comment to explain that it stops `yield_task` working?

Reply:
Amended. I've also hidden `EspWifiPacketBuffer` because it's not public API.